src/Components/Browser.JS/src/Boot.Server.ts

```diff
@@ -6,70 +6,97 @@ import { OutOfProcessRenderBatch } from './Rendering/RenderBatch/OutOfProcessRen
 import { internalFunctions as uriHelperFunctions } from './Services/UriHelper';
 import { renderBatch } from './Rendering/Renderer';
 import { fetchBootConfigAsync, loadEmbeddedResourcesAsync } from './BootCommon';
+import { CircuitHandler } from './Platform/Circuits/CircuitHandler';
 
-let connection : signalR.HubConnection;
+async function boot(reconnect: boolean = false) {
+  const circuitHandlers = new Array<CircuitHandler>();
+  window['Blazor'].addCircuitHandler = (circuitHandler: CircuitHandler) => circuitHandlers.push(circuitHandler);
 
-function boot() {
-  // In the background, start loading the boot config and any embedded resources
-  const embeddedResourcesPromise = fetchBootConfigAsync().then(bootConfig => {
-    return loadEmbeddedResourcesAsync(bootConfig);
-  });
+  await startCicuit();
 
-  connection = new signalR.HubConnectionBuilder()
-    .withUrl('_blazor')
-    .withHubProtocol(new MessagePackHubProtocol())
-    .configureLogging(signalR.LogLevel.Information)
-    .build();
+  async function startCicuit(): Promise<void> {
+    // In the background, start loading the boot config and any embedded resources
+    const embeddedResourcesPromise = fetchBootConfigAsync().then(bootConfig => {
+      return loadEmbeddedResourcesAsync(bootConfig);
+    });
+
+    const initialConnection = await initializeConnection();
+
+    // Ensure any embedded resources have been loaded before starting the app
+    await embeddedResourcesPromise;
+    const circuitId = await initialConnection.invoke<string>(
+      'StartCircuit',
+      uriHelperFunctions.getLocationHref(),
+      uriHelperFunctions.getBaseURI()
+    );
+
+    window['Blazor'].reconnect = async () => {
```

Comment:
TBH I'm not a big fan of overwriting the function instance on each reconnect. That would lead to surprising behavior for anyone who's captured the earlier function instances.

Instead of doing this, consider storing the mutable state (e.g., connection ID) as module-level variables, and then exposing normal fixed functions that operate on those module-level variables.

Reply:
OK, thanks!